File: python/leetcode029.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        k = 1
        if (dividend<0 and divisor>0) or (dividend>0 and divisor<0):
            k = -1
        divisor = abs(divisor)
        dividend = abs(dividend)
        res = 0
        while divisor <= dividend:
            n, nb = 1, divisor

            while dividend >= nb:
                dividend, res = dividend - nb, res + n
                n, nb = n<<1, nb<<1
        if k==1:
            return min(res, 2**31-1)
        else:
            return max(-res, -2**31)

File: leetcodec++/leetcode029.cpp
//
// Created by 黄敏辉 on 2019-07-06.
//

class Solution {
public:
    int divide(int dividend, int divisor) {
        int res = 0;
        int sign = 1;

        if(dividend == INT_MIN  && divisor == -1){
            return INT_MAX;
        }

        if(dividend > 0){
            dividend = -dividend;
            sign = -sign;
        }

        if(divisor > 0){
            divisor = -divisor;
            sign = -sign;
        }

        while(true){
            int tmp = divisor;
            int q = 1;

            if (dividend > tmp){
                return -sign*res;
            }

            while (dividend < tmp){
                res += -q;
                dividend -= tmp;
                if (dividend  >= tmp)
                    break;

                tmp = -((-tmp) << 1);
                q <<= 1;
            }

            if (dividend == tmp){
                res += -q;
                return -sign*res;
            }
        }
    }
};